### ISSUE TITLE
Podman build is very slow compared to docker

### ISSUE THREAD

**Issue by bmaupin** [containers/buildah#1849]
**Description**

I've previously experienced slowness using `podman build` with images that have labels (https://github.com/containers/buildah/issues/1764). But now it seems building any image is very slow.

In case it's relevant, I'm also experiencing an `image not known` error that could be affecting buildah: https://github.com/containers/libpod/issues/3982

Normally I'd wipe ~/.local/share/containers/storage and compare the results with a clean setup but I didn't want to do that this time in case there's something in there that's needed for the other issue. 

**Steps to reproduce the issue:**

Here's my setup:
```
echo 'FROM registry.access.redhat.com/rhoar-nodejs/nodejs-8
WORKDIR /usr/src/app

USER root
RUN chgrp -R 0 . && \
    chmod -R g=u .
USER 1001

COPY package*.json ./
RUN npm ci --only=production

COPY . .

EXPOSE 3000
CMD [ "npm", "start" ]' > Dockerfile

echo "{}" > package-lock.json

echo '{"lockfileVersion": 1}' > package-lock.json
```

```
$ time docker build --no-cache -q .
sha256:87580070c28511b922f556de0924ec92b8867fe6183149ebf19b4a1bdcdaa34c

real	0m14.917s
user	0m0.031s
sys	0m0.021s

$ time docker build -q .
sha256:87580070c28511b922f556de0924ec92b8867fe6183149ebf19b4a1bdcdaa34c

real	0m0.147s
user	0m0.027s
sys	0m0.020s
```

```
$ time podman build --no-cache -q .
02b95eb8d13e124c46cacea3fe59a66da70cb81a3ec91e80be2ddc9cbf5eac0d
1366916962926f8d7f92ab183f8ba567af7855703fcc4b068affd0112f3fcc6c
9c1929bc809d14e6db188ccbde91c2d9420c4792813b944b37f29dc0a209c235
9b393f6380618ab2fbcc5c4ce1315b46b55a656dd23ece360c8cf1360a902917
53fa68f07a2bbf1b287a594a5199ee4c04d65915cb36e68018321bf339fd4dd9
npm info it worked if it ends with ok
npm info using npm@6.4.1
npm info using node@v8.16.1
npm info prepare initializing installer
npm info prepare Done in 0.023s
npm info extractTree Done in 0.003s
npm info updateJson Done in 0.001s
npm info lifecycle undefined@undefined~preinstall: undefined@undefined
npm info lifecycle undefined@undefined~install: undefined@undefined
npm info lifecycle undefined@undefined~postinstall: undefined@undefined
npm info buildTree Done in 0.002s
npm info garbageCollect Done in 0s
npm info lifecycle undefined@undefined~prepublish: undefined@undefined
npm info runScript Done in 0.001s
npm info lifecycle undefined@undefined~prepare: undefined@undefined
npm info runScript Done in 0s
npm info teardown Done in 0s
npm info run-scripts total script time: 0.003s
npm info run-time total run time: 0.032s
added 0 packages in 0.032s
npm timing npm Completed in 649ms
npm info ok 
4c72ca9afede91c9b0e90b0ee33bd888455903281e7c01aed2ec4dd53f581122
cae9f65c8d7ed74c4ae4b79a3f5e85f2ad5cecf6f772780981fc78353aa974df
616ee7f4c6f7422c48ea0673866223a4e8b41714468c3e6bc05b698f4e1f17a3
4e13c4def0df9885e364b140acad223623b91e920ddd18cd2a694779b7d4f37e

real	4m51.123s
user	0m11.320s
sys	0m12.268s

$ time podman build -q .
--> Using cache 9b574591fcab120dfdcf52ecbb6354661605ea19beaadbc37579205073145673
44d5e0c652e668c577cd84e54ee10857709b9911e1374e342bf99b498ed046f0
180f83a3a2c07a53116e6a22a48262381c8178c88f565a855f85067142c3945a
fd3fb6a0fe00f1f536a33c4f4b2d4a42b4c6011bef5c2f6753d3562ac061c4cc
37d7e86ef305e94ab639ef6a9f50bacf828983cbce46bb171053f48fe459fee9
npm info it worked if it ends with ok
npm info using npm@6.4.1
npm info using node@v8.16.1
npm info prepare initializing installer
npm info prepare Done in 0.033s
npm info extractTree Done in 0.003s
npm info updateJson Done in 0.001s
npm info lifecycle undefined@undefined~preinstall: undefined@undefined
npm info lifecycle undefined@undefined~install: undefined@undefined
npm info lifecycle undefined@undefined~postinstall: undefined@undefined
npm info buildTree Done in 0.003s
npm info garbageCollect Done in 0s
npm info lifecycle undefined@undefined~prepublish: undefined@undefined
npm info runScript Done in 0.001s
npm info lifecycle undefined@undefined~prepare: undefined@undefined
npm info runScript Done in 0s
npm info teardown Done in 0s
npm info run-scripts total script time: 0.003s
npm info run-time total run time: 0.042s
added 0 packages in 0.042s
npm timing npm Completed in 741ms
npm info ok 
1c8685182192308e517d085fbdbb5f2e6e83bb6cef91b453c4dac323741f5f8d
9740a1d490dd06edc25cd52a3fd992c7098810d97b5bc2a0a3799e417ea9ea72
f699ec3e38ff680986876008275d81a3ec7adb225a6af62e255e28485cda4c7a
e0052b5acc355383435520e6f5bed8f77097da9ff3c3b61c587f9fb56c245569

real	4m35.198s
user	0m10.698s
sys	0m12.506s

$ time podman build -q .
--> Using cache 9b574591fcab120dfdcf52ecbb6354661605ea19beaadbc37579205073145673
--> Using cache 44d5e0c652e668c577cd84e54ee10857709b9911e1374e342bf99b498ed046f0
--> Using cache 180f83a3a2c07a53116e6a22a48262381c8178c88f565a855f85067142c3945a
--> Using cache fd3fb6a0fe00f1f536a33c4f4b2d4a42b4c6011bef5c2f6753d3562ac061c4cc
--> Using cache 37d7e86ef305e94ab639ef6a9f50bacf828983cbce46bb171053f48fe459fee9
--> Using cache 1c8685182192308e517d085fbdbb5f2e6e83bb6cef91b453c4dac323741f5f8d
--> Using cache 9740a1d490dd06edc25cd52a3fd992c7098810d97b5bc2a0a3799e417ea9ea72
--> Using cache f699ec3e38ff680986876008275d81a3ec7adb225a6af62e255e28485cda4c7a
--> Using cache e0052b5acc355383435520e6f5bed8f77097da9ff3c3b61c587f9fb56c245569

real	2m58.452s
user	0m4.189s
sys	0m8.889s
```

**Describe the results you received:**

`podman build` builds images much slower than docker, with or without a cache. I also had high CPU and IO usage.

**Describe the results you expected:**

I would've expected results comparable to docker.

I also found it interesting that the first time I ran `podman build` without `--no-cache`, it only seemed to use the cache for one layer. I had to run it a second time for it to use the cache for all layers. Docker doesn't seem to exhibit this behaviour.

**Output of `rpm -q buildah` or `apt list buildah`:**

```
$ apt list buildah
Listing... Done
buildah/bionic,now 1.10.1-1~ubuntu18.04~ppa1 amd64 [installed]
```

**Output of `buildah version`:**

```
Version:         1.10.1
Go Version:      go1.10.4
Image Spec:      1.0.1
Runtime Spec:    1.0.1-dev
CNI Spec:        0.4.0
libcni Version:  
Git Commit:      
Built:           Thu Aug  8 16:29:48 2019
OS/Arch:         linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

```
Version:            1.5.1
RemoteAPI Version:  1
Go Version:         go1.10.4
OS/Arch:            linux/amd64
```

**Output of `cat /etc/*release`:**

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.3 LTS"
NAME="Ubuntu"
VERSION="18.04.3 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.3 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

**Output of `uname -a`:**

```
Linux host 5.0.0-27-generic #28~18.04.1-Ubuntu SMP Thu Aug 22 03:00:32 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# storage.conf is the configuration file for all tools
# that share the containers/storage libraries
# See man 5 containers-storage.conf for more information

# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary read-write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# AdditionalImageStores is used to pass paths to additional read-only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers (currently overlay, zfs, vfs, btrfs)
size = ""

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"
```

Thanks!

**Comment by rhatdan** [comment]
Are you sure you are using fuse-overlayfs?

$ podman info

**Comment by rhatdan** [comment]
You are using vfs.

  GraphDriverName: vfs

To get to fuse-ovleray, you need to remove all of your storage and config files and start again.

$ sudo dnf install -y fuse-overlayfs
$ rm -rf ~/.config/containers ~/.local/share/containers
$ podman info 

Now you should be using fuse-overlayfs.

**Comment by scorpionlion** [comment]
Hi, 
I am facing the similar issue where podman rootless build is taking 5 times more than the docker build. Following above steps I also found that the config is using VFS. The OS version I am using is RHEL7.7 and unable to find the package for fuse-ovelayfs in any of 7.7 repos, can you please advise how can I install this package in 7.7?

**Comment by rhatdan** [comment]
Fuse-overlay will not be available until RHEL7.8.
Either move to RHEL8 or use rootfull podman.

**Comment by scorpionlion** [comment]
Okay thanks for getting back. Please can you advise how to run the rootfull podman without prefixing sudo before the command podman build, do we have group similar to docker which we can add user to?

**Comment by rhatdan** [comment]
No podman does not have a daemon, and containers are children of podman.  So you have to run it as root or non-root.  BTW having access to the Docker socket is less secure then running sudo podman.

**Comment by bmaupin** [comment]
@rhatdan

> To get to fuse-ovleray, you need to remove all of your storage and config files and start again.
> 
>  $ sudo dnf install -y fuse-overlayfs

I'm using Ubuntu 18.04. It doesn't look like the fuse-overlayfs package is available until Ubuntu 19.04:

https://packages.ubuntu.com/search?keywords=fuse-overlayfs

**Comment by rhatdan** [comment]
Sorry, not much we can do about that, unless you can get fuse-overlay built for that version of ubuntu. If you run buildah as root, it should be fine.

**Comment by jbertozzi** [comment]
We are also using RHEL7.

The COPY --chown used to run at "normal" speed when running rootless buildah build until a recent update to RHEL7.7. We have run thousands of rootless builds over the last few months without any issue.

We used podman/buildah from rhel-7-server-extras-rpms repository (podman 0.12 / buildah 1.5.2).

I am just trying to understand why it used to work...

**Comment by jbertozzi** [comment]
Just to let you know that downgrading to buildah-1.5-2.gite94b4f9.el7 makes the build fast again on RHEL7.7

I downgraded the whole stack (note sure if it is needed):

yum downgrade containers-common-0.1.31-8.gitb0b750d.el7.x86_64 containernetworking-plugins-0.7.4-1.el7.x86_64 container-selinux-2.77-1.el7_6.noarch podman-0.12.1.2-2.git9551f6b.el7.x86_64 skopeo-0.1.31-8.gitb0b750d.el7.x86_64 buildah-1.5-2.gite94b4f9.el7

**Comment by TomSweeneyRedHat** [comment]
@jbertozzi as you're on RHEL 7.7, fuse-overlayfs is not in the environment and that's why you hit the speed issues.  Downgrading avoids the need for fuse-overlayfs, restricts the usability of rootless containers.  @giuseppe anything else?

**Comment by denysvitali** [comment]
I think that this is related to the fact that the `COPY` instruction, when the `.dockerignore` file is present, copies the files one by one by using [the following snippet](https://github.com/containers/buildah/blob/master/add.go#L303-L339).  
In my case, with a `node_modules` directory, it takes __ages__ because it copies the files one by one:

```
DEBU copyFileWithTar(/home/jenkins/.local/share/containers/storage/overlay/0c718b7bb7789ca658e14c6f800754e7d9e0329e026f1057f7f02e8b55fd49fb/merged/usr/src/app/node_modules/jsdoc/templates/default/static/fonts/OpenSans-BoldItalic-webfont.svg, /home/jenkins/.local/share/containers/storage/overlay/2ca13ffcbc761950b1dd797a04f3bedc699c53f84b2459dafb9601024bb7aeb2/merged/node_modules/jsdoc/templates/default/static/fonts/OpenSans-BoldItalic-webfont.svg)
DEBU error closing OpenSans-BoldItalic-webfont.svg: invalid argument
DEBU copyFileWithTar(/home/jenkins/.local/share/containers/storage/overlay/0c718b7bb7789ca658e14c6f800754e7d9e0329e026f1057f7f02e8b55fd49fb/merged/usr/src/app/node_modules/jsdoc/templates/default/static/fonts/OpenSans-BoldItalic-webfont.woff, /home/jenkins/.local/share/containers/storage/overlay/2ca13ffcbc761950b1dd797a04f3bedc699c53f84b2459dafb9601024bb7aeb2/merged/node_modules/jsdoc/templates/default/static/fonts/OpenSans-BoldItalic-webfont.woff)
DEBU error closing OpenSans-BoldItalic-webfont.woff: invalid argument
DEBU copyFileWithTar(/home/jenkins/.local/share/containers/storage/overlay/0c718b7bb7789ca658e14c6f800754e7d9e0329e026f1057f7f02e8b55fd49fb/merged/usr/src/app/node_modules/jsdoc/templates/default/static/fonts/OpenSans-Italic-webfont.eot, /home/jenkins/.local/share/containers/storage/overlay/2ca13ffcbc761950b1dd797a04f3bedc699c53f84b2459dafb9601024bb7aeb2/merged/node_modules/jsdoc/templates/default/static/fonts/OpenSans-Italic-webfont.eot)
DEBU error closing OpenSans-Italic-webfont.eot: invalid argument
DEBU copyFileWithTar(/home/jenkins/.local/share/containers/storage/overlay/0c718b7bb7789ca658e14c6f800754e7d9e0329e026f1057f7f02e8b55fd49fb/merged/usr/src/app/node_modules/jsdoc/templates/default/static/fonts/OpenSans-Italic-webfont.svg, /home/jenkins/.local/share/containers/storage/overlay/2ca13ffcbc761950b1dd797a04f3bedc699c53f84b2459dafb9601024bb7aeb2/merged/node_modules/jsdoc/templates/default/static/fonts/OpenSans-Italic-webfont.svg)
DEBU error closing OpenSans-Italic-webfont.svg: invalid argument
DEBU copyFileWithTar(/home/jenkins/.local/share/containers/storage/overlay/0c718b7bb7789ca658e14c6f800754e7d9e0329e026f1057f7f02e8b55fd49fb/merged/usr/src/app/node_modules/jsdoc/templates/default/static/fonts/OpenSans-Italic-webfont.woff, /home/jenkins/.local/share/containers/storage/overlay/2ca13ffcbc761950b1dd797a04f3bedc699c53f84b2459dafb9601024bb7aeb2/merged/node_modules/jsdoc/templates/default/static/fonts/OpenSans-Italic-webfont.woff)
```

I'm looking for a solution to this problem, ideally we should tar the file by using an exclusion policy (like `tar --exclude`) instead of excluding the files, creating a file list and `tar`ing the files one by one

**Comment by denysvitali** [comment]
I can confirm that by adding this patch (it is just a workaround that ignores the `.dockerignore`) the build takes a lot less time (but it includes the unwanted files).

```
diff --git a/add.go b/add.go
index b5119e36..032a3cb4 100644
--- a/add.go
+++ b/add.go
@@ -299,6 +299,16 @@ func (b *Builder) addHelper(excludes *fileutils.PatternMatcher, extract bool, de
                                        }
                                }
                                logrus.Debugf("copying[%d] %q to %q", n, esrc+string(os.PathSeparator)+"*", dest+string(os.PathSeparator)+"*")
+                               if excludes != nil {
+                                       logrus.Debugf("excludes= exclusions=%v ", excludes.Exclusions())
+
+                                       for _, p := range excludes.Patterns() {
+                                               logrus.Debugf("pattern=%v", p.String())
+                                               logrus.Debugf("pattern=%v", p.Exclusion())
+                                       }
+
+                                       excludes = nil
+                               }

                                // Copy the whole directory because we do not exclude anything
                                if excludes == nil {
```

And outputs this:

```
level=debug msg="excludes= exclusions=false "
level=debug msg="pattern=.dockerignore"
level=debug msg="pattern=false"
```
  
  
As a workaround, we can maybe implement a deletion after `untar` has been performed: it depends on a case by case basis though. I don't know how Docker handles this kind of copy, but I don't think that `tar`ring single files is a good idea, especially when copying many small files like in the case of a `node_modules` directory

**Comment by TomSweeneyRedHat** [comment]
@denysvitali thanks for the debugging and snooping.  Sounds like you're on the right track.  If you'd like to dive further, please assign this issue to yourself.  If you don't have time, @QiWang19 PTAL when you return.

**Comment by denysvitali** [comment]
@TomSweeneyRedHat Unfortunately I don't have more time to look into this in depth 😓

**Comment by TomSweeneyRedHat** [comment]
@denysvitali no worries, thanks again for the investigation.

**Comment by denysvitali** [comment]
#2072 should fix this problem :D

**Comment by QiWang19** [comment]
@denysvitali Thanks. I'll close this one.

**Comment by ancmikesmullin** [comment]
`podman` and `buildah` are both generally slower at builds at every stage of the build process, AFAICT. I don't have the time to wait for those builds. :(

If comparable-or-greater speed is a priority, show us your benchmarks vs. `docker` on a variety of popular containers using standard hardware like AWS `m5.xlarge`. I'm giving the benefit of the doubt that an optimization pass is planned but just hasn't happened yet, since it's still really early.

**Comment by rhatdan** [comment]
Well we know we have issues with COPY and ADD, which we are working on.  Other then that, we have lots of strategies to go way faster then Docker for certain workloads.

**Comment by peer2peer** [comment]
It's really too slowly. When I build my image with docker, it took 3 minutes but if with podman, about 1 hour. Too much time is spent on "copy blob ...".

My podman info:

`host:
  BuildahVersion: 1.9.0
  Conmon:
    package: podman-1.4.4-4.el7.x86_64
    path: /usr/libexec/podman/conmon
    version: 'conmon version 0.3.0, commit: unknown'
  Distribution:
    distribution: '"rhel"'
    version: "7.2"
  MemFree: 868753408
  MemTotal: 8201007104
  OCIRuntime:
    package: runc-1.0.0-65.rc8.el7.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.1-dev'
  SwapFree: 4294692864
  SwapTotal: 4294963200
  arch: amd64
  cpus: 4
  hostname: ngpe72mgtb26.localdomain
  kernel: 3.10.0-1062.7.1.el7.x86_64
  os: linux
  rootless: false
  uptime: 6h 34m 53.39s (Approximately 0.25 days)
registries:
  blocked: null
  insecure:
  - 10.22.34.164:5000
  search:
  - 10.22.34.164:5000
store:
  ConfigFile: /etc/containers/storage.conf
  ContainerStore:
    number: 6
  GraphDriverName: overlay
  GraphOptions: null
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 79
  RunRoot: /var/run/containers/storage
  VolumePath: /var/lib/containers/storage/volumes`

**Comment by giuseppe** [comment]
> It's really too slowly. When I build my image with docker, it took 3 minutes but if with podman, about 1 hour. Too much time is spent on "copying blob ...".

do you have a reproducer?  Could you share your Dockerfile?

**Comment by peer2peer** [comment]
> > It's really too slowly. When I build my image with docker, it took 3 minutes but if with podman, about 1 hour. Too much time is spent on "copying blob ...".
> 
> do you have a reproducer? Could you share your Dockerfile?

Always.
`FROM rhel7.6:v2

ADD ./myapp /bin/myapp
ADD ./*.so* /lib/
ADD ./mypkg /usr/local/my-pkg

RUN chmod 777 /bin/myapp
WORKDIR /
ENTRYPOINT ["/bin/myapp"]`

**Comment by rhatdan** [comment]
We know we are slow with Huge COPY commands with .dockerignore files.
@nalind is working on a full rewrite of this code.

podman 1.4.4 is an ancient version of podman.  Podman 1.6 should be available on RHEL7 at this point.
podman 1.9.* should be the next release on RHEL8 coming out this summer.

**Comment by rhatdan** [comment]
Fixed in container-selinux v2.134.0

**Comment by lc-thomas** [comment]
Is a full rewrite of the code really needed ? apparently it's much faster with older versions of buildah, is there any reason for that ?

**Comment by nalind** [comment]
It didn't handle .dockerignore correctly, mainly, and the fix for that was bolted on in a way that slowed things down considerably.  That reworking landed in 1.16, though.
Extracting layer contents when using overlay should have improved when https://github.com/containers/storage/pull/631 was merged.
